auth/router.go

```diff
@@ -0,0 +1,150 @@
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/errdefs"
+	"golang.org/x/xerrors"
+
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/docker/docker/api/server/router"
+)
+
+const (
+	issuer = "testdocker"
+)
+
+type authRouter struct {
+	routes []router.Route
+	auth   Auth
+}
+
+// ref. https://docs.docker.com/registry/spec/auth/token/#requesting-a-token
+type TokenResponse struct {
+	Token        string    `json:"token"`
+	AccessToken  string    `json:"access_token"`
+	ExpiresIn    int       `json:"expires_in"`
+	IssuedAt     time.Time `json:"issued_at"`
+	RefreshToken string    `json:"refresh_token"`
+}
+
+type Auth struct {
+	User     string // required
+	Password string // required
+	Secret   string // required
+}
+
+func (a Auth) IsValid() bool {
+	return a.User != "" && a.Password != "" && a.Secret != ""
+}
+
+// NewRouter initializes a new auth router
+func NewRouter(auth Auth) *authRouter {
+	r := &authRouter{
+		auth: auth,
+	}
+	r.initRoutes()
+	return r
+}
+
+// Routes returns the available routes to the image controller
+func (a *authRouter) Routes() []router.Route {
+	return a.routes
+}
+
+// initRoutes initializes the routes in the image router
+func (a *authRouter) initRoutes() {
+	a.routes = []router.Route{
+		// GET
+		router.NewGetRoute("/token", a.tokenHandler), // issue a registry token
+	}
+}
+
+func (a *authRouter) tokenHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	// Authorization: Basic dGVzdDp0ZXN0cGFzcw==
+	authorization := r.Header.Get("Authorization")
+
+	// Basic dGVzdDp0ZXN0cGFzcw==
+	s := strings.Fields(authorization)
+	if len(s) != 2 {
+		return errdefs.Unauthorized(xerrors.New("invalid Authorization header"))
+	}
+	if s[0] != "Basic" {
+		return errdefs.Unauthorized(xerrors.New("'Basic' must be specified"))
+	}
+
+	// dGVzdDp0ZXN0cGFzcw==
+	decoded, err := base64.StdEncoding.DecodeString(s[1])
+	if err != nil {
+		return errdefs.Unauthorized(err)
+	}
+
+	// test:testpass
+	s = strings.Split(string(decoded), ":")
+	if s[0] != a.auth.User || s[1] != a.auth.Password {
+		return errdefs.Unauthorized(xerrors.New("invalid username/password"))
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"iss": issuer,
+	})
+
+	tokenString, err := token.SignedString([]byte(a.auth.Secret))
+	if err != nil {
+		return err
```

Comment:
What do you think about this?
```suggestion
		return errdefs.Unavailable(err)
```

Reply:
Nica catch! Thanks.
https://github.com/aquasecurity/testdocker/pull/2/commits/e9b63fc782c4f77bbd544d6a6642b25ab198b1f4